Introducing Translation Coodinator Role and e2e tests for new CD Admin Dashboard: GSoCM2.2

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #17440 .
2. This PR does the following: 
- Introduces new role of Translation Coordinator.
- Restricted views for contributor-admin-dashboard.

The views will be as follows

        Translation Coordinator: 
- Translation Submitters
- Translation Reviewes


         Question Coordinator:
- Question Submitters
- Question Reviewers


**- End-to-End tests for new contributor admin dashboard.**

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct
[Screencast from 22-09-23 09:01:38 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/3239cba1-d37e-4264-9e11-8481b84a0a63)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
implemented translation coordinator role and handler changes

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix mppy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc_M2.1

kshitij01042002 committed:
first pass by Sean

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
reviewer changes

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
Sean's comments addressed

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
fixed docstring

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
first draft pass

kshitij01042002 committed:
second draft

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
fixed e2e ci error

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 commented:
@kevintab95 @vojtechjelinek @DubeySandeep can you PTAL.

FYI: Oppia bot mentions of adding a new storage model but the storage model was introduced in previous PR and this PR is the branch of that one. This PR doesn't introduces any new storage model.

kshitij01042002 commented:
Hi @Lawful2002 @sagangwee @JayVivarekar can you PTAL. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

Lawful2002 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -46,28 +47,20 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
-    this.TAB_NAME_QUESTION_REVIEWER];
+  CONTRIBUTION_TYPES!: string[];
+  selection!: string;
+
```
Is the extra newline needed here?

kshitij01042002 replied:
Not necessary, fixed it!

Lawful2002 commented on core/tests/webdriverio_utils/ReleaseCoordinatorPage.js:
```diff
@@ -123,6 +123,58 @@ var ReleaseCoordinatorPage = function() {
     return null;
   };
 
+  // Remove this method after the cd_admin_dashboard_new_ui feature flag
```
Create an issue for this and provide the link here. Prefix the comment with `TODO:`.

kshitij01042002 replied:
Done!.

Lawful2002 reviewed (changes requested):
Thanks @kshitij01042002, a few comments.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
reviewers commit

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 commented:
Hi @Lawful2002 made the requested changed, can you PTAL. Thanks

kshitij01042002 committed:
fixed lint errot

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

sagangwee commented:
@kshitij01042002 Can you post screenshots of the UI changes?

kshitij01042002 commented:
@sagangwee updated. PTAL. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

Lawful2002 reviewed (approved):
LGTM for codeowner files.

sagangwee commented:
@Ash-2k3 Can you take a first pass for review?

JayVivarekar reviewed (changes requested):
Thanks @kshitij01042002! Looks good for codeowner files. Can you update the video in the "Proof of changes" section to show what happens when a user without these roles (translation and question coordinator) tries to access the contributor admin page? Ideally, such a user should not be able to access these pages, right?

Ash-2k3 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -46,28 +47,19 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
-    this.TAB_NAME_QUESTION_SUBMITTER,
-    this.TAB_NAME_QUESTION_REVIEWER];
+  CONTRIBUTION_TYPES!: string[];
+  selection!: string;
```
Can you please explain what is this selection variable for ?

kshitij01042002 replied:
This variable is used for changing values in dropdown when the user is in mobile view. You can see it's usage [here](https://github.com/oppia/oppia/pull/18867/files/89df8afc1ed1ead7c8be3046a52e02d3d6944edd#diff-de34ddb8699ce7fec21d168a2c7bb71e330ed07d363b7e9530c0e5f41c3e4a72R55).

Ash-2k3 replied:
Got it thanks, but then isn't selection a bit vague ? It's not conveying what's being selected, probably something like `selectedContributionType` would be more accurate, same for the `updateSelection`.

kshitij01042002 replied:
Sg. I have changed both things please take a look. Thanls.

Ash-2k3 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -76,6 +68,47 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+
+  updateSelection(selection: string): void {
+    this.selection = selection;
+    this.setActiveTab(selection);
```
In updateSelection are we just setting activeTabName ? If yes why not just call `this.setActiveTab` directly instead of creating a new method here, or is there anything I m missing here ?

kshitij01042002 replied:
Actually this function is used in multiple places to update the selected value and the active tab value.

Ash-2k3 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -76,6 +68,47 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+            this.TAB_NAME_QUESTION_REVIEWER];
+        }
+
+        this.updateSelection(this.CONTRIBUTION_TYPES[0]);
```
PTAL at previous comment.

kshitij01042002 replied:
Selected value is updated via dropdown and activeTab value is updated via higher level tabs.

Ash-2k3 reviewed (changes requested):
Thanks @kshitij01042002, left some comments PTAL.

kshitij01042002 commented:
Hi @Ash-2k3 I have replied to all your comments. PTAL.

kshitij01042002 commented:
> Thanks @kshitij01042002! Looks good for codeowner files. Can you update the video in the "Proof of changes" section to show what happens when a user without these roles (translation and question coordinator) tries to access the contributor admin page? Ideally, such a user should not be able to access these pages, right?

Hi @JayVivarekar I have updated video with the case where the user don't have any permission. PTAL again. Thanks for the review.

kshitij01042002 committed:
changed variable names

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 commented:
Hi @Ash-2k3 @JayVivarekar I have addressed all your comments. PTAL again. Thanks for thre review.

Ash-2k3 reviewed (approved):
LGTM, Thanks! Assigning @sagangwee for the second pass.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

JayVivarekar reviewed (approved):
Thanks for the changes @kshitij01042002! LGTM for codeowner files.

vojtechjelinek reviewed (approved):
LGTM!

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+        elem = await elem.editItem(0, 'Dictionary');
+        elem = await elem.editEntry(4, 'List');
+        elem = await elem.addItem('Unicode');
+        // eslint-disable-next-line oppia/e2e-action
```
WHy are we using this flag here?

kshitij01042002 replied:
Here lint check prevents us from directly using `elem.setValue()` it suggests to use `action.setValue()` but if I change it, it doens't seem to set the value correctly and tests fail.

DubeySandeep replied:
Can you please check action.setValue, how it's different any why it's causing this problem?

DubeySandeep replied:
`action.setValue` internally calls `elem.setValue` so it should work, can you please check what's going on? In general, we avoid having flags and use the actions to avoid unexpected flakiness in CI.

kshitij01042002 replied:
I think the reason is elem is a param for the async function and on passing it to `action.setValue` it is loosing context.

DubeySandeep replied:
I'm finding it difficult to understand, it would be great if you could explain it in detail. 

Regarding the `function` in context is `async`, that's because we are using `await` inside, right? The entire `beforeAll` section is in `async` (line 74).

kshitij01042002 replied:
Actually the passing the value of clickable element as `false` while calling action.setValue does the work.

DubeySandeep replied:
Yeahs, so basically we don't want to click the element before setting the value! Thanks for looking into it and fixing it! :)

(We try to avoid using pragmas/flags, in case we do, we write proper documentation/code-comment on why it's needed. Also, developers copy code from one place to another and the pragma/flags gets copied so that was the reason I was trying to get rid of it. I hope you can understand! :) )

kshitij01042002 replied:
Yeah that's understandable and it's always better to follow good coding practices. I am learning new things here. Thanks.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    var explorationEditorSettingsTab = explorationEditorPage.getSettingsTab();
+    await explorationEditorPage.navigateToSettingsTab();
+    let width = (await browser.getWindowSize()).width;
+    if (width < 769) {
```
WHy do we need this logic  here?

kshitij01042002 replied:
I used the existing refrence here. Removed it.

DubeySandeep replied:
I see it in other places as well, should we remove it?

kshitij01042002 replied:
Removed it.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    let copyButton = $('.e2e-test-copy-button');
+    let doneButton = $('.e2e-test-close-rich-text-component-editor');
+    let saveButton = $('.e2e-test-save-button');
+    let textbox = $('.e2e-test-description-box');
```
I'm expecting this to be defined inside utils file ie., mostly the elements are defined in the respective utils file.

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+      'shqip (Albanian)');
+    await contributorDashboardPage.waitForOpportunitiesToLoad();
+    await action.click('Opportunity button', opportunityActionButtonCss);
+    let image = $('.e2e-test-image');
+    await waitFor.visibilityOf(
+      image,
+      'Test image taking too long to appear.');
+    let images = await $$('.e2e-test-image');
+    expect(images.length).toEqual(1);
```
what element  is an `image`? i.,e is it a  button or static image? How does it have the static `e2e-test-image` class?

kshitij01042002 replied:
The image is neither a button nor static image. It is part of card content which is to be translated to a different language. That's the reason it have a static class.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    expect(images.length).toEqual(1);
+
+    // Copy tool should copy image on pressing 'Done'.
+    await waitFor.visibilityOf(
+      copyButton, 'Copy button taking too long to appear');
+    await action.click('Copy button', copyButton);
+    await action.click('Image', images[0]);
+    await action.setValue(
+      'Set Image description', textbox, 'An example description');
+    await action.click('Done', doneButton);
+    await action.click('Save', saveButton);
+    await users.logout();
```
Maybe  define this in utils file as  copyElementWithClassName(elementDescription, elementClassName) and use it here*

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    await contributorDashboardPage.waitForOpportunitiesToLoad();
+    await contributorDashboardPage.selectReviewLanguage('shqip (Albanian)');
+
+    if (width < 700) {
+      let opportunityItemSelector = function() {
+        return $$('.e2e-test-opportunity-list-item');
+      };
+      let opportunity = (await opportunityItemSelector())[0];
+      await action.click('Opportunity Item', opportunity);
+    } else {
+      await contributorDashboardPage.clickOpportunityActionButton(
+        'Chapter 1', 'Topic 0 for contribution - Story Title');
+    }
```
WHy are we having the width check here? (Don't we have control over width when running e2e tests?)

If we want to test behavior for different screens, can't we have different describe/it blocks for each scenario?

kshitij01042002 replied:
These are setUp steps are we are not testing this. I needed to populate data for stats to be displayed in cd admin dashboard, so most of this is legacy code which I reused to generate data.

DubeySandeep replied:
It's okay to use existing code but at your own responsibility to make sure you understand what's going on and only use the required steps. Can we remove things that are not required for this test setup?

kshitij01042002 replied:
Removed it.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    if (width < 700) {
+      let opportunityItemSelector = function() {
+        return $$('.e2e-test-opportunity-list-item');
+      };
+      let opportunity = (await opportunityItemSelector())[0];
+      await action.click('Opportunity Item', opportunity);
```
maybe define this in the utils file as clickFiestOpportunityInSmallScreenView?

Also, why are we selecting the random first item where as for bigger screen selecting a specific item?

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+        '[Image]', 'Topic 0 for contribution / Story Title / Chapter 1');
+    }
+
+    var acceptButton = $('.e2e-test-translation-accept-button');
+    await action.click('Translation accept button', acceptButton);
```
Define thsi in utils file*

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,351 @@
+    await users.login(QUESTION_COORDINATOR_EMAIL);
+    await contributorDashboardAdminPage.get();
+
+    var questionSubmitterTab = $('.e2e-test-question-submitters-tab');
+    var questionReviewerTab = $('.e2e-test-question-reviewers-tab');
+
+    await action.click('Question Submitter tab', questionSubmitterTab);
```
Ditto define these in utils files, follow this pattern for all other palces*

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
   };
+
+  this.checkstatslistelements = async function(number) {
+    await waitFor.presenceOf(statsTable, 'stats table was not visible');
```
```suggestion
    await waitFor.presenceOf(statsTable, 'Stats table is not visible');
```

kshitij01042002 replied:
Done. PTAL.

DubeySandeep replied:
I think the initial capital `S` is still not died (not sure!), can you please check and fix it?

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
     expect(await contributionRights.getText()).toBe('Allowed');
   };
+
+  this.checkstatslistelements = async function(number) {
```
name of the function should be camelCase*

DubeySandeep replied:
Maybe call it expectStatsElementCountToBe

kshitij01042002 replied:
Updated. Thanks for nit.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
+    expect(statsListItems.length).toBe(number);
+  };
+
+  this.checkexpandablestats = async function() {
```
DItto*

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
     expect(await contributionRights.getText()).toBe('Allowed');
   };
+
+  this.checkstatslistelements = async function(number) {
```
The `number` here creates confusion with the type `number` in the type script. Maybe use elementsCount*

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
+  };
+
+  this.checkexpandablestats = async function() {
+    await waitFor.presenceOf(statsTable, 'stats table was not visible');
```
why precenceOf here and visibiltyOf elesewhere?

kshitij01042002 replied:
Actually I was doing some debugging with `presenceOf` and forgot to change it to `visibiltyOf`. Changed. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -177,6 +182,21 @@ var ContributorDashboardAdminPage = function() {
+    expect(statsListItems.length).toBe(number);
+  };
+
+  this.checkexpandablestats = async function() {
```
what are we checking here?

kshitij01042002 replied:
We are checking if the row is expanded on click or not.

DubeySandeep replied:
so maybe rename it to expectStatsRowsAreEpanded().

Note: The `expect` prefix is used for testing blocks.

kshitij01042002 replied:
Sg. Updated.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -251,6 +267,23 @@ var ContributorDashboardPage = function() {
+
+  this.navigateToMyContributionTab = async function() {
+    await action.click(
+      'Submit question tab button', myContributionTabButton);
```
Incorrect description*

kshitij01042002 replied:
Changed. Thanks for the nit.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -52,6 +56,18 @@ var ContributorDashboardPage = function() {
+  var selectorContainer = $('.e2e-test-language-selector');
+
+  var _openLanguageSelector = async function() {
+    await action.click('Test Language Selector Container', selectorContainer);
```
WHy are we having `Test` prefix here?

kshitij01042002 replied:
I wrote it in the sense of Unit tests and forgot it was e2e tests. Sorry for that. Updated/

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -52,6 +56,18 @@ var ContributorDashboardPage = function() {
+    await _openLanguageSelector();
+    var selectorOption = selectorContainer.$(
+      `.e2e-test-language-selector-option=${language}`);
+    await action.click('Test Language Selector Option', selectorOption);
```
Ditto*

DubeySandeep replied:
Add `language` value in the decsrition, helps with debugging i.e, the error message will clearly show which language was attempted for selection

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -52,6 +56,18 @@ var ContributorDashboardPage = function() {
+  var selectorContainer = $('.e2e-test-language-selector');
+
+  var _openLanguageSelector = async function() {
+    await action.click('Test Language Selector Container', selectorContainer);
```
Add a check to ensure the langaugeSelector is opened.

kshitij01042002 replied:
Done. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -251,6 +267,23 @@ var ContributorDashboardPage = function() {
+  this.selectTranslationReviewTab = async function() {
+    var reviewTabs = await reviewTabsSelector();
+    await action.click(
+      'Open translation review opportunity', reviewTabs[1]
```
The description should be the name of the button/element. Maybe "Review tab button"

kshitij01042002 replied:
Updated. PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -251,6 +267,23 @@ var ContributorDashboardPage = function() {
+  };
+
+  this.selectTranslationReviewTab = async function() {
+    var reviewTabs = await reviewTabsSelector();
```
Do we have multiple review tabs button in view?

kshitij01042002 replied:
Yeah, we have multiple review tabs. I think review buttons would be a more suitable name for it.

DubeySandeep commented on core/tests/webdriverio_utils/ReleaseCoordinatorPage.js:
```diff
@@ -123,6 +123,58 @@ var ReleaseCoordinatorPage = function() {
+  this.enableFeatureForTest = async function(featureElement) {
+    await action.click(
+      'Add feature rule button',
+      featureElement
+        .$(addFeatureRuleButtonLocator)
```
```suggestion
      featureElement.$(addFeatureRuleButtonLocator)
```

kshitij01042002 replied:
Done.

DubeySandeep reviewed (changes requested):
@kshitij01042002 Thanks for the PR, I've left a few comments on e2e file changes, PTAL!

kshitij01042002 committed:
sandeeps's review comments

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 commented:
Hi @DubeySandeep thanks for the review. This was my first time writing e2e tests. Your comments really helped to learn new things about writing e2e tests, will keep in mind next time. Thanks.

I have addressed all your comments. PTAL.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -76,6 +68,47 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+  }
+
+  ngOnInit(): void {
+    let userInfoPromise = this.userService.getUserInfoAsync();
```
```suggestion
    this.userService.getUserInfoAsync().then(userInfo => {
```

kshitij01042002 replied:
Updated.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -76,6 +68,47 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+    let userInfoPromise = this.userService.getUserInfoAsync();
+    userInfoPromise.then(userInfo => {
+      const username = userInfo.getUsername();
+      if (username) {
```
Flip the condition using a guard clause to reduce nesting and increase readability.
```suggestion
      if (username === null) {
        return;
      }
     ...
```

kshitij01042002 replied:
Sg. Done.

sagangwee reviewed (approved):
@kshitij01042002 LGTM—just two suggestions.

kshitij01042002 committed:
suggestions by Sagang

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 commented:
Thanks @sagangwee for the review and suggestions. Implemented them.

kshitij01042002 committed:
lint errors

kshitij01042002 committed:
test coverage

DubeySandeep commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -27,6 +28,115 @@ describe('Contributor dashboard Admin page', () => {
   let contributorDashboardAdminStatsBackendApiService: (
     ContributorDashboardAdminStatsBackendApiService);
+  let userService: UserService;
+  let translationCoordinatorInfo = {
```
is there any issue with creating the UserInfo object here? (The existing code is not maintainable i.e, if someone ad/removes a function from UserInfo class they will have to modify this place unexpectedly.)

Ditto below*

kshitij01042002 replied:
If I create a UserInfo object with only attributes needed, it throws error of missing attributes.

DubeySandeep replied:
Are you sure? (Note: By UserInfo obejct I mean `new UserInfo(...)`.)

Can you please share the log, I can help you fi it if needed?

kshitij01042002 replied:
Yeah that does the work. I might have been doing something wrong. Thanks for the help.

DubeySandeep commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -76,6 +68,47 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+        this.CONTRIBUTION_TYPES = [this.TAB_NAME_QUESTION_SUBMITTER,
+          this.TAB_NAME_QUESTION_REVIEWER];
+      }
+      if (this.isQuestionCoordinator && this.isTranslationCoordinator) {
+        this.CONTRIBUTION_TYPES = [this.TAB_NAME_TRANSLATION_SUBMITTER,
+          this.TAB_NAME_TRANSLATION_REVIEWER,
+          this.TAB_NAME_QUESTION_SUBMITTER,
+          this.TAB_NAME_QUESTION_REVIEWER];
+      }
```
Do we need this block? Considering the above two if black will result in the same value for the given situation.

kshitij01042002 replied:
The above blocks will initialize the array with only 2 values for each true condition.

DubeySandeep replied:
Can't we simplfy it to:
```
CONTRIBUTION_TYPES = [];
if(A) CONTRIBUTION_TYPES.push(a,b)
if(B) CONTRIBUTION_TYPES.push(c,d)
```

This way we won't have to duplicate check off `if(A&B)`, right?

kshitij01042002 replied:
Yeah. Sg. Implemented.

DubeySandeep commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -46,28 +47,19 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
-    this.TAB_NAME_QUESTION_REVIEWER];
+  CONTRIBUTION_TYPES!: string[];
+  selectedContributionType!: string;
+  isQuestionCoordinator!: boolean;
+  isTranslationCoordinator!: boolean;
```
Do we need this in the template? IF not, is there any specific reason for binding it to `object/this`?

kshitij01042002 replied:
Yes we need it in the template to render some components based on these flags.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,322 @@
+  const USER_EMAILS = ['user0@contributor.com', 'user1@contributor.com'];
+  const QUESTION_ADMIN_EMAIL = 'user@contributor.com';
+
+  var releaseCoordinatorPage = null;
+  var adminPage = null;
+  var contributorDashboardAdminPage = null;
+  let contributorDashboardPage = null;
+  let topicsAndSkillsDashboardPage = null;
```
Any specific reason for using var and let for different places here?

kshitij01042002 replied:
No speicfic reason, used it alternatively

DubeySandeep replied:
Let''s use `let` in all the place to avoid confusion?

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,322 @@
+    await explorationEditorMainTab.setInteraction('EndExploration');
+    var explorationEditorSettingsTab = explorationEditorPage.getSettingsTab();
+    await explorationEditorPage.navigateToSettingsTab();
+    let width = (await browser.getWindowSize()).width;
```
Do we need `width`?

kshitij01042002 replied:
Removed it.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,322 @@
+    var explorationEditorSettingsTab = explorationEditorPage.getSettingsTab();
+    await explorationEditorPage.navigateToSettingsTab();
+    let width = (await browser.getWindowSize()).width;
+    var basicSettings = $('.e2e-test-settings-container');
+    await action.click('Basic Settings', basicSettings);
```
Why do we need to do this?

kshitij01042002 replied:
We are setting up attributes of exploration via settings tab like title, category, objective, etc

DubeySandeep replied:
Ah, I mean why do we need to click `basicSettings`? Is it any form of button or dropdown?

kshitij01042002 replied:
It's a tab.

DubeySandeep replied:
SO are we clicking it to open the tab? or is it already open? (Basically, I'm unable to understand what this click is doing and why do we need it?)

kshitij01042002 replied:
![image](https://github.com/oppia/oppia/assets/91470808/13f5ab9f-be24-4d44-b4a7-cc21814723bc)

I think this image will give you a better idea.

DubeySandeep commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -0,0 +1,322 @@
+    await contributorDashboardPage.waitForOpportunitiesToLoad();
+    await contributorDashboardPage.selectReviewLanguage('shqip (Albanian)');
+
+    if (width < 700) {
```
Do we need these logics? if not please remove it from here and elsewhere*

(Asking this as it adds discrepancy between tests i.e, there can be unexpected behaviour in the tests. In general, we avoid adding logic in the tests.)

kshitij01042002 replied:
Removed it.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    let copyButton = $('.e2e-test-copy-button');
+    let doneButton = $('.e2e-test-close-rich-text-component-editor');
+    let saveButton = $('.e2e-test-save-button');
+    let textbox = $('.e2e-test-description-box');
```
Define these on the top (maybe)?

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    let saveButton = $('.e2e-test-save-button');
+    let textbox = $('.e2e-test-description-box');
+
+    // Copy tool should copy image on pressing 'Done'.
```
This comment doesn't seems relevent for the next line*

kshitij01042002 replied:
Removed it.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    // Copy tool should copy image on pressing 'Done'.
+    await waitFor.visibilityOf(
+      copyButton, 'Copy button taking too long to appear');
+    await action.click('Copy button', copyButton);
```
`await action.click('Copy button', copyButton);` CAn you please check whether action.click internally take cares of visibilityOf, if yes, please remove the duplicate check on line 131

kshitij01042002 replied:
Done. Thanks for pointing out.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await action.click(elementDescription, elementClassName);
+    await action.setValue(
+      'Set Image description', textbox, 'An example description');
+    await action.click('Done', doneButton);
```
```suggestion
    await action.click('Copy Done button', doneButton);
```

DubeySandeep replied:
Is it resolved?

kshitij01042002 replied:
Updated it before merging.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await action.setValue(
+      'Set Image description', textbox, 'An example description');
+    await action.click('Done', doneButton);
+    await action.click('Save', saveButton);
```
```suggestion
    await action.click('Save button', saveButton);
```

DubeySandeep replied:
is thsi resolved?

kshitij01042002 replied:
Updated it before merging.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await users.logout();
+  };
+
+  this.clickFirstOpportunityInSmallScreenView = async function() {
```
If we are going to remove width, then remove this if it's not needed.

kshitij01042002 replied:
Removed.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+  };
+
+  this.acceptTranslation = async function() {
+    var acceptButton = $('.e2e-test-translation-accept-button');
```
Define all elements on top*

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+  };
+
+  this.navigateToquestionSubmitterTab = async function() {
+    var questionSubmitterTab = $('.e2e-test-question-submitters-tab');
```
Ditto here and elsewhere*

kshitij01042002 replied:
Done

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await action.click('Question Submitter tab', questionSubmitterTab);
+  };
+
+  this.navigateToquestionReviewerTab = async function() {
```
```suggestion
  this.navigateToQuestionReviewerTab = async function() {
```

DubeySandeep replied:
is this resolved?

kshitij01042002 replied:
Updated it before merging.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await action.click('Question Reviewer tab', questionReviewerTab);
+  };
+
+  this.navigateTotranslationSubmitterTab = async function() {
```
```suggestion
  this.navigateToTranslationSubmitterTab = async function() {
```

DubeySandeep replied:
Is this resolved?

kshitij01042002 replied:
Updated it before merging.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -114,6 +120,58 @@ var ContributorDashboardAdminPage = function() {
+    await action.click('Translation Submitter tab', translationSubmitterTab);
+  };
+
+  this.navigateTotranslationReviewerTab = async function() {
```
```suggestion
  this.navigateToTranslationReviewerTab = async function() {
```

DubeySandeep replied:
is this resolved?

kshitij01042002 replied:
Updated it before merging.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -52,6 +56,22 @@ var ContributorDashboardPage = function() {
+    await _openLanguageSelector();
+    var selectorOption = selectorContainer.$(
+      `.e2e-test-language-selector-option=${language}`);
+    await action.click(language + ' Language Selector Option', selectorOption);
```
```suggestion
    await action.click(`Language ${language} option`, selectorOption);
```

DubeySandeep replied:
is this resolved?

kshitij01042002 replied:
Yes. have updated it before merging.

DubeySandeep reviewed (approved):
LGTM, left a few comments, PTAL!

Also, feel free o merge the PR once the remaining comments are resolved!

kshitij01042002 committed:
reviewers comments

DubeySandeep commented:
@kshitij01042002 Can you please reply to all the open comments? Also, please take a pass on this thread: https://github.com/oppia/oppia/pull/18867#discussion_r1339543632

kshitij01042002 commented:
Hi @DubeySandeep I have created a new PR addressing the pending comments #18951 . can you PTAL? Thanks.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardPage.js:
```diff
@@ -52,6 +56,22 @@ var ContributorDashboardPage = function() {
+  };
+
+  var _selectLanguage = async function(language) {
+    await _openLanguageSelector();
+    var selectorOption = selectorContainer.$(
+      `.e2e-test-language-selector-option=${language}`);
+    await action.click(`Language ${language} option`, selectorOption);
```
Can you please check whether we can use `action.select`?
